### PR TITLE
Feature/rome support

### PR DESCRIPTION
**Pull request by miscampbell**
### Short description 📝

Adds support to retrieve Carthage dependencies using [Rome](https://github.com/tmspzz/Rome). This framework enables you to retrieve the dependencies from a cached repository such as AWS S3.

### Solution 📦

The solution implemented follows the same pattern added for UpCarthage. I have only implemented the method rome download as this is the only function I see users wanting to use under `tuist up`

### Implementation 👩‍💻👨‍💻

Adds support for the command `rome download --platforms [Specified In Code] --cache-prefix [Specified In Code]
We also check if the Rome tap is available before when running the UpRome command

- [ ] Run Unit Tests from UpRomeTests
- [ ] Add .rome() to your Setup.swift file to test

### CODE REVIEW

File: Sources/TuistLoader/Up/Rome.swift [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import Foundation
+import TSCBasic
+import TuistCore
+import TuistSupport
+
+/// Protocol that defines an interface to interact with a local Rome setup.
+protocol Romeaging {
```
Comment (pepibumur) [inline Sources/TuistLoader/Up/Rome.swift:7]: Nice name 😜

File: Sources/TuistLoader/Up/UpRome.swift [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    /// Returns true when the command doesn't need to be run.
+    ///
+    /// - Parameters
+    ///   - projectPath: Path to the directory that contains the project manifest.
+    /// - Returns: True if the command doesn't need to be run.
+    /// - Throws: An error if the check fails.
+    override func isMet(projectPath: AbsolutePath) throws -> Bool {
+        if try !upHomebrew.isMet(projectPath: projectPath) { return false }
+        return true
```
Comment (pepibumur) [inline Sources/TuistLoader/Up/UpRome.swift:67]: Is there a way to know if Rome has already pulled those pre-compiled dependencies? If so, we should add the logic here instead of returning true. Otherwise, every time someone does `tuist up` they'll end up downloading the frameworks even if they already exist locally.

File: Sources/TuistLoader/Up/UpRome.swift [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        /// Installing Rome
+        if try !upHomebrew.isMet(projectPath: projectPath) {
+            try upHomebrew.meet(projectPath: projectPath)
+        }
+
+        /// Downloading Rome dependencies.
+        try rome.download(platforms: platforms, cachePrefix: cachePrefix)
+    }
+
+    func whatever() {}
```
Comment (pepibumur) [inline Sources/TuistLoader/Up/UpRome.swift:85]: I think we can delete this one :)

File: Tests/TuistLoaderTests/Up/UpRomeTests.swift [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    func test_init() throws {
+        let temporaryPath = try self.temporaryPath()
+        let json = JSON(["platforms": JSON.array([JSON.string("ios")]), "cachePrefix": "Swift_5_1"])
+        let got = try UpRome(dictionary: json, projectPath: temporaryPath)
+        XCTAssertEqual(got.platforms, [.iOS])
+        XCTAssertEqual(got.cachePrefix, "Swift_5_1")
+    }
+
+    func test_isMet_when_rome_is_not_met() throws {
```
Comment (pepibumur) [inline Tests/TuistLoaderTests/Up/UpRomeTests.swift:46]: I'd recommend being consistent with how we structure the body of the tests in the rest of the codebase:

```swift
// Given
...

// When
...

// Then
...

```
Using those comments makes the tests easier to read and reason about.

File: Tests/TuistLoaderTests/Up/UpRomeTests.swift [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    var platforms: [Platform]!
+    var cachePrefix: String!
+    var upHomebrew: MockUp!
+    var rome: MockRome!
+    var subject: UpRome!
+
+    override func setUp() {
+        super.setUp()
+        platforms = [.iOS]
+        cachePrefix = "Swift_1_2"
```
Comment (mollyIV) [inline Tests/TuistLoaderTests/Up/UpRomeTests.swift:20]: Maybe we could also add tests for an empty / nil `platforms` param and no `cache prefix`? 🤔

File: Sources/TuistLoader/Up/UpRome.swift (hunk 2 above)
Comment (mollyIV) [reply Sources/TuistLoader/Up/UpRome.swift:67]: According to the [rome's documentation](https://github.com/tmspzz/Rome#usage), a `list` command looks promising:

```bash
rome list --help
...
Lists frameworks in the cache and reports cache misses/hits, according to the
local Cartfile.resolved. Ignores dSYMs.

--missing                List frameworks missing from the cache. Ignores dSYMs
--present                List frameworks present in the cache. Ignores dSYMs.
```

Maybe we could try to leverage one of these flags? 🤔

Comment (miscampbell) [reply Sources/TuistLoader/Up/UpRome.swift:67]: Hi @mollyIV thank you for looking into this. I am just about to push a change up now to use rome list that will check for any missing frameworks in advance.

File: Sources/TuistLoader/Up/UpRome.swift (hunk 3 above)
Comment (miscampbell) [reply Sources/TuistLoader/Up/UpRome.swift:85]: Thank @pepibumur this has been removed

File: Sources/TuistLoader/Up/Rome.swift [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    /// Retrieves the dependencies in the given directory.
+    ///
+    /// - Parameters:
+    ///   - path: Directory where the Carthage dependencies are defined.
```
[the remark this answers is not in the document]
Comment (miscampbell) [reply Sources/TuistLoader/Up/Rome.swift:11]: Good spot @mollyIV this is now removed :) I spotted this when I was adding the rome list support

**Comment by miscampbell** [comment]
Hi @pepibumur @mollyIV thank you for taking the time to review this PR and your valuable feedback. I have just committed all suggested changes please let me know any further tweaks are needed for improvements.

Should any further changes be required please let me know.
